File: test/erl/src/test_client.erl
%%
%% Licensed to the Apache Software Foundation (ASF) under one
%% or more contributor license agreements. See the NOTICE file
%% distributed with this work for additional information
%% regarding copyright ownership. The ASF licenses this file
%% to you under the Apache License, Version 2.0 (the
%% "License"); you may not use this file except in compliance
%% with the License. You may obtain a copy of the License at
%%
%%   http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing,
%% software distributed under the License is distributed on an
%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
%% KIND, either express or implied. See the License for the
%% specific language governing permissions and limitations
%% under the License.
%%

-module(test_client).

-export([start/0, start/1]).

-include("thriftTest_types.hrl").

-record(options, {port = 9090,
                  client_opts = []}).

parse_args(Args) -> parse_args(Args, #options{}).
parse_args([], Opts) -> Opts;
parse_args([Head | Rest], Opts) ->
    NewOpts =
        case catch list_to_integer(Head) of
            Port when is_integer(Port) ->
                Opts#options{port = Port};
            _Else ->
                case Head of
                    "framed" ->
                        Opts#options{client_opts = [{framed, true} | Opts#options.client_opts]};
                    "" ->
                        Opts;
                    _Else ->
                        erlang:error({bad_arg, Head})
                end
        end,
    parse_args(Rest, NewOpts).


start() -> start([]).
start(Args) ->
  #options{port = Port, client_opts = ClientOpts} = parse_args(Args),
  {ok, Client0} = thrift_client_util:new(
    "127.0.0.1", Port, thriftTest_thrift, ClientOpts),

  DemoXtruct = #xtruct{
    string_thing = <<"Zero">>,
    byte_thing = 1,
    i32_thing = 9128361,
    i64_thing = 9223372036854775807},

  DemoNest = #xtruct2{
    byte_thing = 7,
    struct_thing = DemoXtruct,
    % Note that we don't set i32_thing, it will come back as undefined
    % from the Python server, but 0 from the C++ server, since it is not
    % optional
    i32_thing = 2},

  % Is it safe to match these things?
  DemoDict = dict:from_list([ {Key, Key-10} || Key <- lists:seq(0,10) ]),
  DemoSet = sets:from_list([ Key || Key <- lists:seq(-3,3) ]),

  %DemoInsane = #insanity{
  %  userMap = dict:from_list([{?thriftTest_FIVE, 5000}]),
  %  xtructs = [#xtruct{ string_thing = <<"Truck">>, byte_thing = 8, i32_thing = 8, i64_thing = 8}]},

  {Client01, {ok, ok}} = thrift_client:call(Client0, testVoid, []),

  {Client02, {ok, <<"Test">>}}      = thrift_client:call(Client01, testString, ["Test"]),
  {Client03, {ok, <<"Test">>}}      = thrift_client:call(Client02, testString, [<<"Test">>]),
  {Client04, {ok, 63}}              = thrift_client:call(Client03, testByte, [63]),
  {Client05, {ok, -1}}              = thrift_client:call(Client04, testI32, [-1]),
  {Client06, {ok, 0}}               = thrift_client:call(Client05, testI32, [0]),
  {Client07, {ok, -34359738368}}    = thrift_client:call(Client06, testI64, [-34359738368]),
  {Client08, {ok, -5.2098523}}      = thrift_client:call(Client07, testDouble, [-5.2098523]),
  {Client09, {ok, DemoXtruct}}      = thrift_client:call(Client08, testStruct, [DemoXtruct]),
  {Client10, {ok, DemoNest}}        = thrift_client:call(Client09, testNest, [DemoNest]),
  {Client11, {ok, DemoDict}}        = thrift_client:call(Client10, testMap, [DemoDict]),
  {Client12, {ok, DemoSet}}         = thrift_client:call(Client11, testSet, [DemoSet]),
  {Client13, {ok, [-1,2,3]}}        = thrift_client:call(Client12, testList, [[-1,2,3]]),
  {Client14, {ok, 1}}               = thrift_client:call(Client13, testEnum, [?thriftTest_Numberz_ONE]),
  {Client15, {ok, 309858235082523}} = thrift_client:call(Client14, testTypedef, [309858235082523]),

  % No python implementation, but works with C++ and Erlang.
  %{Client16, {ok, InsaneResult}}    = thrift_client:call(Client15, testInsanity, [DemoInsane]),
  %io:format("~p~n", [InsaneResult]),
  Client16 = Client15,

  {Client17, {ok, #xtruct{string_thing = <<"Message">>}}} =
    thrift_client:call(Client16, testMultiException, ["Safe", "Message"]),

  Client18 =
    try
      {ClientS1, Result1} = thrift_client:call(Client17, testMultiException, ["Xception", "Message"]),
      io:format("Unexpected return! ~p~n", [Result1]),
      ClientS1
    catch
      throw:{ClientS2, {exception, ExnS1 = #xception{}}} ->
        #xception{errorCode = 1001, message = <<"This is an Xception">>} = ExnS1,
        ClientS2;
      throw:{ClientS2, {exception, _ExnS1 = #xception2{}}} ->
        io:format("Wrong exception type!~n", []),
        ClientS2
    end,

  Client19 =
    try
      {ClientS3, Result2} = thrift_client:call(Client18, testMultiException, ["Xception2", "Message"]),
      io:format("Unexpected return! ~p~n", [Result2]),
      ClientS3
    catch
      throw:{ClientS4, {exception, _ExnS2 = #xception{}}} ->
        io:format("Wrong exception type!~n", []),
        ClientS4;
      throw:{ClientS4, {exception, ExnS2 = #xception2{}}} ->
        #xception2{errorCode = 2002,
                   struct_thing = #xtruct{
                     string_thing = <<"This is an Xception2">>}} = ExnS2,
        ClientS4
    end,

  thrift_client:close(Client19).

File: lib/csharp/test/ThriftTest/TestClient.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Threading;
using Thrift.Collections;
using Thrift.Protocol;
using Thrift.Transport;
using Thrift.Test;

namespace Test
{
	public class TestClient
	{
		private static int numIterations = 1;

		public static void Execute(string[] args)
		{
			try
			{
				string host = "localhost";
				int port = 9090;
				string url = null;
				int numThreads = 1;
				bool buffered = false, framed = false;

				try
				{
					for (int i = 0; i < args.Length; i++)
					{
						if (args[i] == "-h")
						{
							string[] hostport = args[++i].Split(':');
							host = hostport[0];
							if (hostport.Length > 1)
							{
								port = Convert.ToInt32(hostport[1]);
							}
						}
						else if (args[i] == "-u")
						{
							url = args[++i];
						}
						else if (args[i] == "-n")
						{
							numIterations = Convert.ToInt32(args[++i]);
						}
						else if (args[i] == "-b" || args[i] == "-buffered")
						{
							buffered = true;
							Console.WriteLine("Using buffered sockets");
						}
						else if (args[i] == "-f" || args[i] == "-framed")
						{
							framed = true;
							Console.WriteLine("Using framed transport");
						}
						else if (args[i] == "-t")
						{
							numThreads = Convert.ToInt32(args[++i]);
						}
					}
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}



				//issue tests on separate threads simultaneously
				Thread[] threads = new Thread[numThreads];
				DateTime start = DateTime.Now;
				for (int test = 0; test < numThreads; test++)
				{
					Thread t = new Thread(new ParameterizedThreadStart(ClientThread));
					threads[test] = t;
					if (url == null)
					{
						TTransport trans = new TSocket(host, port);
						if (buffered)
							trans = new TBufferedTransport(trans as TStreamTransport);
						if (framed)
							trans = new TFramedTransport(trans);
							
						t.Start(trans);
					}
					else
					{
						THttpClient http = new THttpClient(new Uri(url));
						t.Start(http);
					}
				}

				for (int test = 0; test < numThreads; test++)
				{
					threads[test].Join();
				}
				Console.Write("Total time: " + (DateTime.Now - start));
			}
			catch (Exception outerEx)
			{
				Console.WriteLine(outerEx.Message + " ST: " + outerEx.StackTrace);
			}

			Console.WriteLine();
			Console.WriteLine();
		}

		public static void ClientThread(object obj)
		{
			TTransport transport = (TTransport)obj;
			for (int i = 0; i < numIterations; i++)
			{
				ClientTest(transport);
			}
			transport.Close();
		}

		public static void ClientTest(TTransport transport)
		{
			TBinaryProtocol binaryProtocol = new TBinaryProtocol(transport);

			ThriftTest.Client client = new ThriftTest.Client(binaryProtocol);
			try
			{
				if (!transport.IsOpen)
				{
					transport.Open();
				}
			}
			catch (TTransportException ttx)
			{
				Console.WriteLine("Connect failed: " + ttx.Message);
				return;
			}

			long start = DateTime.Now.ToFileTime();

			Console.Write("testVoid()");
			client.testVoid();
			Console.WriteLine(" = void");

			Console.Write("testString(\"Test\")");
			string s = client.testString("Test");
			Console.WriteLine(" = \"" + s + "\"");

			Console.Write("testByte(1)");
			sbyte i8 = client.testByte((sbyte)1);
			Console.WriteLine(" = " + i8);

			Console.Write("testI32(-1)");
			int i32 = client.testI32(-1);
			Console.WriteLine(" = " + i32);

			Console.Write("testI64(-34359738368)");
			long i64 = client.testI64(-34359738368);
			Console.WriteLine(" = " + i64);

			Console.Write("testDouble(5.325098235)");
			double dub = client.testDouble(5.325098235);
			Console.WriteLine(" = " + dub);

			Console.Write("testStruct({\"Zero\", 1, -3, -5})");
			Xtruct o = new Xtruct();
			o.String_thing = "Zero";
			o.Byte_thing = (sbyte)1;
			o.I32_thing = -3;
			o.I64_thing = -5;
			Xtruct i = client.testStruct(o);
			Console.WriteLine(" = {\"" + i.String_thing + "\", " + i.Byte_thing + ", " + i.I32_thing + ", " + i.I64_thing + "}");

			Console.Write("testNest({1, {\"Zero\", 1, -3, -5}, 5})");
			Xtruct2 o2 = new Xtruct2();
			o2.Byte_thing = (sbyte)1;
			o2.Struct_thing = o;
			o2.I32_thing = 5;
			Xtruct2 i2 = client.testNest(o2);
			i = i2.Struct_thing;
			Console.WriteLine(" = {" + i2.Byte_thing + ", {\"" + i.String_thing + "\", " + i.Byte_thing + ", " + i.I32_thing + ", " + i.I64_thing + "}, " + i2.I32_thing + "}");

			Dictionary<int, int> mapout = new Dictionary<int, int>();
			for (int j = 0; j < 5; j++)
			{
				mapout[j] = j - 10;
			}
			Console.Write("testMap({");
			bool first = true;
			foreach (int key in mapout.Keys)
			{
				if (first)
				{
					first = false;
				}
				else
				{
					Console.Write(", ");
				}
				Console.Write(key + " => " + mapout[key]);
			}
			Console.Write("})");

			Dictionary<int, int> mapin = client.testMap(mapout);

			Console.Write(" = {");
			first = true;
			foreach (int key in mapin.Keys)
			{
				if (first)
				{
					first = false;
				}
				else
				{
					Console.Write(", ");
				}
				Console.Write(key + " => " + mapin[key]);
			}
			Console.WriteLine("}");

			List<int> listout = new List<int>();
			for (int j = -2; j < 3; j++)
			{
				listout.Add(j);
			}
			Console.Write("testList({");
			first = true;
			foreach (int j in listout)
			{
				if (first)
				{
					first = false;
				}
				else
				{
					Console.Write(", ");
				}
				Console.Write(j);
			}
			Console.Write("})");

			List<int> listin = client.testList(listout);

			Console.Write(" = {");
			first = true;
			foreach (int j in listin)
			{
				if (first)
				{
					first = false;
				}
				else
				{
					Console.Write(", ");
				}
				Console.Write(j);
			}
			Console.WriteLine("}");

			//set
			THashSet<int> setout = new THashSet<int>();
			for (int j = -2; j < 3; j++)
			{
				setout.Add(j);
			}
			Console.Write("testSet({");
			first = true;
			foreach (int j in setout)
			{
				if (first)
				{
					first = false;
				}
				else
				{
					Console.Write(", ");
				}
				Console.Write(j);
			}
			Console.Write("})");

			THashSet<int> setin = client.testSet(setout);

			Console.Write(" = {");
			first = true;
			foreach (int j in setin)
			{
				if (first)
				{
					first = false;
				}
				else
				{
					Console.Write(", ");
				}
				Console.Write(j);
			}
			Console.WriteLine("}");


			Console.Write("testEnum(ONE)");
			Numberz ret = client.testEnum(Numberz.ONE);
			Console.WriteLine(" = " + ret);

			Console.Write("testEnum(TWO)");
			ret = client.testEnum(Numberz.TWO);
			Console.WriteLine(" = " + ret);

			Console.Write("testEnum(THREE)");
			ret = client.testEnum(Numberz.THREE);
			Console.WriteLine(" = " + ret);

			Console.Write("testEnum(FIVE)");
			ret = client.testEnum(Numberz.FIVE);
			Console.WriteLine(" = " + ret);

			Console.Write("testEnum(EIGHT)");
			ret = client.testEnum(Numberz.EIGHT);
			Console.WriteLine(" = " + ret);

			Console.Write("testTypedef(309858235082523)");
			long uid = client.testTypedef(309858235082523L);
			Console.WriteLine(" = " + uid);

			Console.Write("testMapMap(1)");
			Dictionary<int, Dictionary<int, int>> mm = client.testMapMap(1);
			Console.Write(" = {");
			foreach (int key in mm.Keys)
			{
				Console.Write(key + " => {");
				Dictionary<int, int> m2 = mm[key];
				foreach (int k2 in m2.Keys)
				{
					Console.Write(k2 + " => " + m2[k2] + ", ");
				}
				Console.Write("}, ");
			}
			Console.WriteLine("}");

			Insanity insane = new Insanity();
			insane.UserMap = new Dictionary<Numberz, long>();
			insane.UserMap[Numberz.FIVE] = 5000L;
			Xtruct truck = new Xtruct();
			truck.String_thing = "Truck";
			truck.Byte_thing = (sbyte)8;
			truck.I32_thing = 8;
			truck.I64_thing = 8;
			insane.Xtructs = new List<Xtruct>();
			insane.Xtructs.Add(truck);
			Console.Write("testInsanity()");
			Dictionary<long, Dictionary<Numberz, Insanity>> whoa = client.testInsanity(insane);
			Console.Write(" = {");
			foreach (long key in whoa.Keys)
			{
				Dictionary<Numberz, Insanity> val = whoa[key];
				Console.Write(key + " => {");

				foreach (Numberz k2 in val.Keys)
				{
					Insanity v2 = val[k2];

					Console.Write(k2 + " => {");
					Dictionary<Numberz, long> userMap = v2.UserMap;

					Console.Write("{");
					if (userMap != null)
					{
						foreach (Numberz k3 in userMap.Keys)
						{
							Console.Write(k3 + " => " + userMap[k3] + ", ");
						}
					}
					else
					{
						Console.Write("null");
					}
					Console.Write("}, ");

					List<Xtruct> xtructs = v2.Xtructs;

					Console.Write("{");
					if (xtructs != null)
					{
						foreach (Xtruct x in xtructs)
						{
							Console.Write("{\"" + x.String_thing + "\", " + x.Byte_thing + ", " + x.I32_thing + ", " + x.I32_thing + "}, ");
						}
					}
					else
					{
						Console.Write("null");
					}
					Console.Write("}");

					Console.Write("}, ");
				}
				Console.Write("}, ");
			}
			Console.WriteLine("}");


			sbyte arg0 = 1;
			int arg1 = 2;
			long arg2 = long.MaxValue;
			Dictionary<short, string> multiDict = new Dictionary<short, string>();
			multiDict[1] = "one";
			Numberz arg4 = Numberz.FIVE;
			long arg5 = 5000000;
			Console.Write("Test Multi(" + arg0 + "," + arg1 + "," + arg2 + "," + multiDict + "," + arg4 + "," + arg5 + ")");
			Xtruct multiResponse = client.testMulti(arg0, arg1, arg2, multiDict, arg4, arg5);
			Console.Write(" = Xtruct(byte_thing:" + multiResponse.Byte_thing + ",String_thing:" + multiResponse.String_thing
						+ ",i32_thing:" + multiResponse.I32_thing + ",i64_thing:" + multiResponse.I64_thing + ")\n");

			Console.WriteLine("Test Oneway(1)");
			client.testOneway(1);

			Console.Write("Test Calltime()");
			var startt = DateTime.UtcNow;
			for ( int k=0; k<1000; ++k )
				client.testVoid();
			Console.WriteLine(" = " + (DateTime.UtcNow - startt).TotalSeconds.ToString() + " ms a testVoid() call" );
		}
	}
}
